### PR TITLE
Exporters Performance Optimizations

### LINKED ISSUE #6673: Issue with export of csv file on data grid
## Description  
Is there any data limitation on export to csv? we are trying to export around 500000 records from grid and it gets hanged. totally unresponsive. What is best practice to follow for export from grid? and any other workaround for this issue? 
 
 * igniteui-angular version: 
 * browser:  Chrome

## Steps to reproduce  
Export large data from data grid

## Result  
screens becomes unresponsive

## Expected result  
it should export all data to csv

### PR DESCRIPTION
**Pull request by DiyanDimitrov**
Closes #6673

### Additional information (check all that apply):
 - [x] Bug fix
 - [ ] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [ ] All relevant tags have been applied to this PR
 - [ ] This PR includes unit tests covering all the new code ([test guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Test-implementation-guidelines-for-Ignite-UI-for-Angular))
 - [ ] This PR includes API docs for newly added methods/properties ([api docs guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Documentation-Guidelines))
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [ ] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes ([migrations guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Update-Migrations))
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

File: projects/igniteui-angular/src/lib/services/excel/excel-files.ts [hunk 1]
```diff
@@ -131,6 +136,105 @@ export class WorksheetFile implements IExcelFile {
...
+            this.processDataRecordsAsync(worksheetData, (rows) => {
+                sheetData += rows;
+                sheetData += '</sheetData>';
+                done(cols, sheetData);
+            });
+        }
+    }
+
+    private processDataRecordsAsync(worksheetData: WorksheetData, done: (rows: string) => void) {
+        let dataRecords = '';
```
Comment (DiyanDimitrov) [inline projects/igniteui-angular/src/lib/services/excel/excel-files.ts:202]: Replace with array.

File: projects/igniteui-angular/src/lib/services/excel/excel-exporter.ts [hunk 2]
```diff
@@ -55,14 +56,33 @@ export class IgxExcelExporterService extends IgxBaseExporter {
...
+            const folderInstance = ExcelElementsFactory.getExcelFolder(childFolder);
+            const zipFolder = zip.folder(folderInstance.folderName);
+            IgxExcelExporterService.populateFolderAsync(folderInstance, zipFolder, worksheetData, done);
+        }
+
+        for (const childFile of folder.childFiles(worksheetData)) {
+            const fileInstance = ExcelElementsFactory.getExcelFile(childFile);
+            if (fileInstance instanceof WorksheetFile) {
+                (fileInstance as WorksheetFile).writeElementAsync(zip, worksheetData, () => {
+                    done();
```
Comment (DiyanDimitrov) [inline projects/igniteui-angular/src/lib/services/excel/excel-exporter.ts:83]: We should wait all childFiles for complete before calling done().

File: projects/igniteui-angular/ng-package.json [hunk 3]
```diff
@@ -8,6 +8,7 @@
   "whitelistedNonPeerDependencies": [
       "@types/hammerjs",
       "@types/jszip",
+      "core-js",
```
Comment (DiyanDimitrov) [inline projects/igniteui-angular/ng-package.json:11]:
```suggestion
      "core-js-pure",
```

File: projects/igniteui-angular/ng-package.prod.json [hunk 4]
```diff
@@ -10,6 +10,7 @@
   "whitelistedNonPeerDependencies": [
       "@types/hammerjs",
       "@types/jszip",
+      "core-js",
```
Comment (DiyanDimitrov) [inline projects/igniteui-angular/ng-package.prod.json:13]:
```suggestion
      "core-js-pure",
```

File: projects/igniteui-angular/src/lib/services/csv/char-separated-value-data.ts [hunk 5]
```diff
@@ -73,6 +93,18 @@ export class CharSeparatedValueData {
         return dataRecords;
     }
 
+    private processDataRecordsAsync(currentData, keys, escapeChars, done: (result: string) => void) {
+        let dataRecords = '';
+        yieldingLoop(currentData.length, 1000,
+            (i) => {
+                const row = currentData[i];
+                dataRecords += this.processRecord(row, keys, escapeChars);
```
Comment (DiyanDimitrov) [inline projects/igniteui-angular/src/lib/services/csv/char-separated-value-data.ts:101]: Replace string concatenation with array join.

File: projects/igniteui-angular/src/lib/services/excel/worksheet-data-dictionary.ts [hunk 6]
```diff
@@ -24,18 +23,20 @@ export class WorksheetDataDictionary {
 
     public stringsCount: number;
 
-    constructor(columnCount: number, columnWidth: number) {
+    // TODO Switch to fixed column width since auto-calculation takes a lot of time
```
Comment (igdmdimitrov) [inline projects/igniteui-angular/src/lib/services/excel/worksheet-data-dictionary.ts:26]: Remove TODO

File: projects/igniteui-angular/src/lib/services/excel/worksheet-data.ts [hunk 7]
```diff
@@ -11,8 +11,8 @@ export class WorksheetData {
     private _isSpecialData: boolean;
 
     constructor(private _data: any[], public options: IgxExcelExporterOptions, public indexOfLastPinnedColumn,
-                public sort: any, public isTreeGridData = false) {
-        this.initializeData();
+                public sort: any, public isTreeGridData = false, public columnWidths: number[]) {
```
Comment (igdmdimitrov) [inline projects/igniteui-angular/src/lib/services/excel/worksheet-data.ts:14]: Change to private _columnWidths and remove parameter from initializeData method.

File: src/main.ts [hunk 8]
```diff
@@ -1,3 +1,5 @@
+// TODO this is fixing the JSZip perf
+// import 'setimmediate';
```
Comment (igdmdimitrov) [inline src/main.ts:2]: Remove commented lines

File: src/app/grid-performance/grid-performance.sample.html [hunk 9]
```diff
@@ -1,6 +1,6 @@
 <div style="height: 100vh; width: 100wh">
     <span> Fixed Size Rows</span>
-    <igx-grid #grid1 [showToolbar]="true" [rowHeight]="20" displayDensity="compact" [columnHiding]="true" [data]="localData" [rowSelection]="selectionMode" [allowFiltering]="true" [filterMode]="'excelStyleFilter'">
+    <igx-grid #grid1 [showToolbar]="true" [rowHeight]="20" displayDensity="compact" [columnHiding]="true" [data]="localData" [rowSelection]="selectionMode" [allowFiltering]="true" [filterMode]="'excelStyleFilter'" [exportCsv]="true" [exportExcel]="true" exportText="Export" exportExcelText="Export to Excel" exportCsvText="Export to CSV">
```
Comment (igdmdimitrov) [inline src/app/grid-performance/grid-performance.sample.html:3]: Remove the changes made to the sample

**Pull request merged**